src/lem/circuit.rs

```diff
@@ -0,0 +1,927 @@
+                                    match $slot {
+                                        SlotType::Hash2 => &g.preallocated_hash2_slots[slots_count.consume_hash2()],
+                                        SlotType::Hash3 => &g.preallocated_hash3_slots[slots_count.consume_hash3()],
+                                        SlotType::Hash4 => &g.preallocated_hash4_slots[slots_count.consume_hash4()],
+                                    };
+
+                                // Add the implication constraint for the image
+                                implies_equal(
+                                    &mut cs.namespace(|| {
+                                        format!("implies equal for {}'s hash (LEMOP {:?})", $img, &op)
+                                    }),
+                                    &concrete_path,
+                                    $allocated_img.hash(),
+                                    &preallocated_img,
+                                )?;
+
+                                // For each component of the preimage, add implication constraints
+                                // for its tag and hash
+                                for (i, allocated_ptr) in $allocated_preimg.iter().enumerate() {
+                                    let name = $preimg[i].name();
+                                    let ptr_idx = 2 * i;
+                                    implies_equal(
+                                        &mut cs.namespace(|| {
+                                            format!("implies equal for {name}'s tag (LEMOP {:?}, pos {i})", &op)
+                                        }),
+                                        &concrete_path,
+                                        allocated_ptr.tag(),
+                                        &preallocated_preimg[ptr_idx], // tag index
+                                    )?;
+                                    implies_equal(
+                                        &mut cs.namespace(|| {
+                                            format!(
+                                                "implies equal for {name}'s hash (LEMOP {:?}, pos {i})",
+                                                &op
+                                            )
+                                        }),
+                                        &concrete_path,
+                                        allocated_ptr.hash(),
+                                        &preallocated_preimg[ptr_idx + 1], // hash index
+                                    )?;
+                                }
+                            };
+                        }
+                        macro_rules! hash_helper {
+                            ( $img: expr, $tag: expr, $preimg: expr, $slot: expr ) => {
+                                // Allocate image. Its hash field is constrained by `constrain_slot!`
+                                let allocated_img = LEM::allocate_ptr(
```

Comment:
I believe this is an unnecessary allocation because the img should already have been allocated in the slot.